### PR TITLE
Added zynq7000-xgpio driver

### PR DESCRIPTION
**Pull request by agkaminski**
JIRA: PP-37

## Description

Driver for AXI GPIO Xilinx IP

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (zynq7000-zedboard).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [x] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: gpio/zynq7000-xgpio/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# zynq7000-xgpio
```
Comment (gerard5) [inline gpio/zynq7000-xgpio/README.md:1]: I think the better place for documentation is the `phoenix-rtos-doc` repo.

File: gpio/zynq7000-xgpio/xgpio.c [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+	return ret;
+}
+
+
+static int gpio_writePin(int channel, int pin, uint32_t val)
+{
+	int ret = -EINVAL;
+
+	if (channel >= 0 && channel <= 1 && pin >= 0 && pin <= 31) {
+		ret = gpio_writePort(channel, val << pin, 1 << pin);
```
Comment (gerard5) [inline gpio/zynq7000-xgpio/xgpio.c:152]:
```suggestion
		ret = gpio_writePort(channel, val << pin, 1U << pin);
```

File: gpio/zynq7000-xgpio/xgpio.c [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+	}
+}
+
+
+static void pwm_usage(const char *progname)
+{
+	printf("Usage: %s [OPTIONS]\n", progname);
+	printf("Options:\n");
+	printf("\t-b base_addr - base address of XGPIO IP\n");
+	printf("\t-p priority  - sever thread priority (default %d)\n", PRIORITY);
```
Comment (gerard5) [inline gpio/zynq7000-xgpio/xgpio.c:280]: `sever` -> `server`.

File: gpio/zynq7000-xgpio/xgpio.c [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+#define GPIO1_DIR  1
+#define GPIO2_DATA 2
+#define GPIO2_DIR  3
+
+
+typedef struct {
+	int channel;
+	int isDir;
+	int isPin;
+	int pin;
```
[the remark this answers is not in the document]
Comment (agkaminski) [reply gpio/zynq7000-xgpio/xgpio.c:39]: unsigned should be really used only when there's a reason for it. It's no in this case - I need plain a integer.

File: gpio/zynq7000-xgpio/xgpio.c [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+} oid_info_t;
+
+
+static struct {
+	oid_t oid;
```
[the remark this answers is not in the document]
Comment (agkaminski) [reply gpio/zynq7000-xgpio/xgpio.c:44]: I avoid casting oid_t * -> void * -> oid_t *, which is a MISRA violation we would need to justify.

File: gpio/zynq7000-xgpio/README.md (hunk 1 above)
Comment (agkaminski) [reply gpio/zynq7000-xgpio/README.md:1]: Where? We don't have specific drivers documentation in phoenix-rtos-doc repository: https://github.com/phoenix-rtos/phoenix-rtos-doc/tree/master/devices.

Also, this isn't really a documentation, it's just a brief usage info - it's a README ;)

File: gpio/zynq7000-xgpio/xgpio.c [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+#define GPIO2_DATA 2
+#define GPIO2_DIR  3
+
+
+typedef struct {
+	int channel;
+	int isDir;
+	int isPin;
+	int pin;
+} oid_info_t;
```
[the remark this answers is not in the document]
Comment (agkaminski) [reply gpio/zynq7000-xgpio/xgpio.c]: I wanted to convey that this struct represents decoded oid. gpio_info_t doesn't really communicate that. Changed anyway

File: gpio/zynq7000-xgpio/README.md (hunk 1 above)
Comment (agkaminski) [reply gpio/zynq7000-xgpio/README.md:1]: Place will be made and READMEs will be moved in the near future

**Pull request merged**